### ISSUE TITLE
SQLITE_CORRUPT - Mac and iOS not syncing

### ISSUE THREAD

**Issue by amandamcg** [laurent22/joplin#5153]
## Environment

Joplin 2.1.7 (prod, darwin)

Client ID: fd585dd87ccd4d55b2bb8d418c2e87e5
Sync Version: 2
Profile Version: 39
Keychain Supported: Yes

I am running a lot of plugins!

Revision: 04c4015
Platform: Mac OS
OS specifics: 11.3.1

Syncing to Webdav

--- 

iOS - iPhone 14.6 
Joplin version 12.1.1

---

iPad & Terminal
These versions are likely out of date, I haven't opened it in some time although I noticed Homebrew updated the terminal app recently. 

---

## Steps to reproduce
This morning, I noticed on mobile that a note from yesterday was conflicted with itself, which was weird because I only had edited it on mobile. This uncovered a bunch of other issues, as follows:

I hit sync manually on mobile and briefly saw an SQL error when it completed. I unfortunately didn't write down what it was. I hit sync again and it reported that it synced "successfully". 

I opened my desktop app and noticed that no notes from mobile had synced since 6/25 and desktop had not synced to mobile since 6/22. 

I manually ran a sync on desktop (which was reporting that all was well before that). I then got this error: 

Completed: 07/02/21 9:32 AM (9s)
Last error: Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT items.`id`,items.`parent_id`,items.`title`,items.`body`,items.`created_time`,items.`updated_time`,items.`is_conflict`,items.`latitude`,items.`longitude`,items.`altitude`,items.`author`,items.`source_url`,items.`is_todo`,items.`todo_due`,items.`todo_completed`,items.`source`,items.`source_application`,items.`application_data`,items.`order`,items.`user_created_time`,items.`user_updated_time`,items.`encryption_cipher_text`

Now, when I hit sync or it syncs automatically I see that error again. Mobile has never again reported an SQL error after the one time I noticed it. It just says it has synced and the time.

When running the sync check, both apps report that Synchronization configuration appears to be correct and that there are no errors. Mac reports that `Note: 2699/2710` have been synced, but doesn't list any errors. 

I am able to create a new test note that persists when I open and close the app. 

I have exported my desktop notes as a JEX, I am not sure how to export the mobile notes, so I have manually backed up the unsynced notes separately. 

## Describe what you expected to happen

If SQL was corrupted, why did it not show as an error until I manually synced desktop? Bigger issue is understanding how to get back into a good state.

## Logfile
/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/Logger.js:188 09:51:31: ResourceService::indexNoteResources: Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT id, title, body, encryption_applied FROM notes WHERE id IN (<6 note IDs>): 
    at DatabaseDriverNode.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database-driver-node.js:21)
    at JoplinDatabase.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:34)
    at JoplinDatabase.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:124)
    at Generator.throw (<anonymous>)
    at rejected (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:6)

/Applications/Joplin.app/Contents/Resources/app.asar/app.js:212 Uncaught (in promise) TypeError: In reducer: Cannot read property 'key' of null Action: {"type":"MAIN_LAYOUT_SET_ITEM_PROP","itemKey":"plugin-view-joplin.plugin.ambrt.backlinksToNote-backlinksPanel","propName":"visible","propValue":true}
    at /Applications/Joplin.app/Contents/Resources/app.asar/app.js:212
    at Object.iterateItems [as default] (/Applications/Joplin.app/Contents/Resources/app.asar/gui/ResizableLayout/utils/iterateItems.js:6)
    at /Applications/Joplin.app/Contents/Resources/app.asar/app.js:211
    at Immer.produce (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/immer/dist/immer.cjs.development.js:748)
    at reducer (/Applications/Joplin.app/Contents/Resources/app.asar/app.js:210)
    at dispatch (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/node_modules/redux/lib/createStore.js:178)
    at Application.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/BaseApplication.js:453)
    at Generator.next (<anonymous>)
    at /Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/BaseApplication.js:8
    at new Promise (<anonymous>)

09:51:33: RevisionService: RevisionService::collectRevisions: Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT * FROM notes WHERE is_conflict = 0 AND encryption_applied = 0 AND id IN (<same 6 note IDs>): 
    at DatabaseDriverNode.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database-driver-node.js:21)
    at JoplinDatabase.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:34)
    at JoplinDatabase.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:124)
    at Generator.throw (<anonymous>)
    at rejected (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:6)

09:51:36: Error: Error: SQLITE_CORRUPT: database disk image is malformed: 
					SELECT items.`id`,items.`parent_id`,items.`title`,items.`body`,items.`created_time`,items.`updated_time`,items.`is_conflict`,items.`latitude`,items.`longitude`,items.`altitude`,items.`author`,items.`source_url`,items.`is_todo`,items.`todo_due`,items.`todo_completed`,items.`source`,items.`source_application`,items.`application_data`,items.`order`,items.`user_created_time`,items.`user_updated_time`,items.`encryption_cipher_text`,items.`encryption_applied`,items.`markup_language`,items.`is_shared`,items.`share_id`,items.`conflict_original_id`,`sync_time` FROM `notes` items
					JOIN sync_items s ON s.item_id = items.id
					WHERE sync_target = 6
					AND (s.sync_time < items.updated_time OR force_sync = 1)
					AND s.sync_disabled = 0
					AND is_conflict = 0 AND encryption_applied = 0
					ORDER BY items.updated_time DESC
					LIMIT 89
				: 
    at DatabaseDriverNode.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database-driver-node.js:21)
    at JoplinDatabase.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:34)
    at JoplinDatabase.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:124)
    at Generator.throw (<anonymous>)
    at rejected (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:6)

/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/Logger.js:188 09:56:43: ResourceService::indexNoteResources: Error: Error: SQLITE_CORRUPT: database disk image is malformed: SELECT id, title, body, encryption_applied FROM notes WHERE id IN (<different 6 note IDs>): 
    at DatabaseDriverNode.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database-driver-node.js:21)
    at JoplinDatabase.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:34)
    at JoplinDatabase.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:124)
    at Generator.throw (<anonymous>)
    at rejected (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:6)

/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/Logger.js:188 09:56:44: Error: Error: SQLITE_CORRUPT: database disk image is malformed: 
					SELECT items.`id`,items.`parent_id`,items.`title`,items.`body`,items.`created_time`,items.`updated_time`,items.`is_conflict`,items.`latitude`,items.`longitude`,items.`altitude`,items.`author`,items.`source_url`,items.`is_todo`,items.`todo_due`,items.`todo_completed`,items.`source`,items.`source_application`,items.`application_data`,items.`order`,items.`user_created_time`,items.`user_updated_time`,items.`encryption_cipher_text`,items.`encryption_applied`,items.`markup_language`,items.`is_shared`,items.`share_id`,items.`conflict_original_id`,`sync_time` FROM `notes` items
					JOIN sync_items s ON s.item_id = items.id
					WHERE sync_target = 6
					AND (s.sync_time < items.updated_time OR force_sync = 1)
					AND s.sync_disabled = 0
					AND is_conflict = 0 AND encryption_applied = 0
					ORDER BY items.updated_time DESC
					LIMIT 89
				: 
    at DatabaseDriverNode.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database-driver-node.js:21)
    at JoplinDatabase.sqliteErrorToJsError (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:34)
    at JoplinDatabase.<anonymous> (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:124)
    at Generator.throw (<anonymous>)
    at rejected (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/database.js:6)

**Comment by amandamcg** [comment]
Trying to dig into this further. 

When I create a new test note I get a new error
`10:11:27: Could not get lat/long for note <NOTE ID>:  FetchError {name: "FetchError", message: "request to https://freegeoip.app/json/ failed, reason: connect ECONNREFUSED 0.0.0.0:443", type: "system", errno: "ECONNREFUSED", code: "ECONNREFUSED", …}`


I have geo location turned off for Joplin at a Mac System Level and had turned off geo location in the app, or so I thought, but now looking at Settings, this is checked again:

- [x] Save geo-location with notes

I am almost positive that I have turned that checkbox off previously, and I am not sure when it would have been turned back on (but it's possible I toggled it at some point). 

Not sure if this could be causing the issue? I am wondering if something changed in a recent version regarding geo? I installed 2.1.7 on 6/28 and the one before that (sorry don't remember number) on 6/27.

**Comment by laurent22** [comment]
Unfortunately once a database is corrupt there's not much to do other than restore from backup. For the SQL error in iOS could you check the log, from the config screen? You should be able to display errors only and normally the SQL error should be there.

**Comment by laurent22** [comment]
And the geolocation error is unrelated - it might just be that the service we use is down.

**Comment by amandamcg** [comment]
> For the SQL error in iOS could you check the log, from the config screen? You should be able to display errors only and normally the SQL error should be there.
I see 4 errors, all "Lock has expired" errors over the past week or so. No SQL error. 


> Unfortunately once a database is corrupt there's not much to do other than restore from backup
Ok, I will start this process tomorrow unless there is any help to you to keep it in this state for debugging.

**Comment by laurent22** [comment]
Thanks but once it's corrupt it's not possible to retrieve useful info from it unfortunately.

Normally the app itself cannot corrupt the database as SQLite is quite resilient so when it happens it's due to external factors.

Did you by any chance point the CLI and/or desktop app to the same profile (using the `--profile` flag)? That could be a reason. Another possible reason is to backup and restore a while profile directory while the application is still running. Otherwise more rarely it can be due to a faulty drive - in that case, just to be sure you might want to run a disk checking utility.

**Comment by amandamcg** [comment]
> Did you by any chance point the CLI and/or desktop app to the same profile (using the `--profile` flag)?

I didn't modify any flags manually, but I occasionally open the app in dev mode using the option from the menu to test out theme things. Not sure if that is relevant. 

> Another possible reason is to backup and restore a while profile directory while the application is still running.

Hmm, no, haven't ever restored the app before.

> Otherwise more rarely it can be due to a faulty drive - in that case, just to be sure you might want to run a disk checking utility.

Thanks for the tip, I ran some tests and nothing seems off, but this is an older machine, so I'll double check my backups just in case. 

---

I'll close this issue for now and start working on restoring. Thanks for taking a look.

**Comment by amandamcg** [comment]
Ok, I am now back up and syncing again. Wanted to follow up with a few notes, just in case it's useful. Feel free to ignore.

What I learned from restoring:
- The note that seems to be the first that didn't sync is one that i have edited multiple times with an external editor. Might be a clue to what went wrong. Maybe there was a collision or crash while saving this note back to the DB?
- Importing Joplin from a RAW directory duplicated tags and didn't put things back in the same notebook (everything goes to `Imported`). I need to figure out the correct way to backup and restore everything. In the end I used an old .zip I had made of joplin-desktop (I found one from before the SQL error started appearing) and reimported just the files between the zip and now. Once I let the app sync again, it turned out most of these (30/45 notes) were on the cloud and resynced to the correct places. I'm still a little concerned about the state of things, but I have lots of backups at different points now, so I will move forward. I would reimport from scratch, but I don't want to lose all of my notebooks.
- There is something wrong with the application layout feature that is causing a crash, I am going to open a new bug with the details, but I realize this could be happening because of how I recovered things.
